### PR TITLE
Support for AudioVM

### PR DESCRIPTION
**Pull request by fepitre**
QubesOS/qubes-issues#1590

### CODE REVIEW

File: qubes/ext/audio.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    def on_property_del(self, subject, event, name, oldvalue=None):
+        newvalue = getattr(subject, 'audiovm', None)
+        self.on_property_set(subject, event, name, newvalue, oldvalue)
+
+    @qubes.ext.handler('property-set:audiovm')
+    def on_property_set(self, subject, event, name, newvalue, oldvalue=None):
+        # pylint: disable=unused-argument,no-self-use
+
+        # Clean other 'audiovm-XXX' tags.
+        # gui-agent can connect to only one domain
```
Comment (marmarek) [inline qubes/ext/audio.py]: pulseaudio agent AKA module-vchan-sink

Comment (fepitre) [reply qubes/ext/audio.py]: Ok thank you. I pushed force with this correction.

File: qubes/ext/audio.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            raise qubes.exc.QubesVMError(
+                self, 'There are running VMs using this VM as AudioVM: '
+                      '{}'.format(', '.join(vm.name for vm in attached_vms)))
+
+    @qubes.ext.handler('domain-init', 'domain-load')
+    def on_domain_init_load(self, vm, event):
+        if getattr(vm, 'audiovm', None):
+            if 'audiovm-' + vm.audiovm.name not in list(vm.tags):
+                vm.fire_event('property-set:audiovm',
+                              name='audiovm', newvalue=vm.audiovm)
```
Comment (marmarek) [inline qubes/ext/audio.py]: This should rather call `self.on_property_set` directly. There may be other handlers for `property-set:audiovm` that may get confused with spurious event.

File: qubes/ext/audio.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        newvalue = getattr(subject, 'audiovm', None)
+        self.on_property_set(subject, event, name, newvalue, oldvalue)
+
+    @qubes.ext.handler('property-set:audiovm')
+    def on_property_set(self, subject, event, name, newvalue, oldvalue=None):
+        # Clean other 'audiovm-XXX' tags.
+        # pulseaudio agent (module-vchan-sink) can connect to only one domain
+        tags_list = list(subject.tags)
+        for tag in tags_list:
+            if 'audiovm-' in tag:
```
Comment (marmarek) [inline qubes/ext/audio.py]: Should be `tag.startswith('audiovm-')`

File: qubes/ext/gui.py [hunk 4]
```diff
@@ -21,20 +21,43 @@
...
         vm.run_service('qubes.SetGuiMode',
                        input=('SEAMLESS'
                               if vm.features.get('gui-seamless', False)
                               else 'FULLSCREEN'))
 
+    @qubes.ext.handler('domain-init', 'domain-load')
+    def on_domain_init_load(self, vm, event):
+        if getattr(vm, 'guivm', None):
+            if 'guivm-' + vm.guivm.name not in list(vm.tags):
+                vm.fire_event('property-set:guivm',
```
Comment (marmarek) [inline qubes/ext/gui.py]: Here too, it should call `self.on_property_set` directly.

File: qubes/ext/audio.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                        domain.is_running()]
+        if attached_vms and not kwargs.get('force', False):
+            raise qubes.exc.QubesVMError(
+                self, 'There are running VMs using this VM as AudioVM: '
+                      '{}'.format(', '.join(vm.name for vm in attached_vms)))
+
+    @qubes.ext.handler('domain-init', 'domain-load')
+    def on_domain_init_load(self, vm, event):
+        if getattr(vm, 'audiovm', None):
+            if 'audiovm-' + vm.audiovm.name not in list(vm.tags):
```
Comment (marmarek) [inline qubes/ext/audio.py]: You don't need to convert tags into a list, it should be enough to use `not in vm.tags`.
It is necessary in few other places, if you add/remove them while iterating.

File: qubes/ext/gui.py [hunk 6]
```diff
@@ -26,15 +26,35 @@
...
     def send_gui_mode(vm):
         vm.run_service('qubes.SetGuiMode',
                        input=('SEAMLESS'
                               if vm.features.get('gui-seamless', False)
                               else 'FULLSCREEN'))
 
+    @qubes.ext.handler('domain-init', 'domain-load')
+    def on_domain_init_load(self, vm, event):
+        if getattr(vm, 'guivm', None):
+            if 'guivm-' + vm.guivm.name not in list(vm.tags):
```
Comment (marmarek) [inline qubes/ext/gui.py]: `list()` unnecessary here too.

File: qubes/tests/app.py [hunk 7]
```diff
@@ -608,6 +608,47 @@ class MyTestHolder(qubes.tests.TestEmitter, qubes.PropertyHolder):
...
+        holder.default_audiovm = 'sys-audio'
+        self.assertEqual(holder.default_audiovm, 'sys-audio')
+        self.assertIsNotNone(self.app.default_audiovm)
+        self.assertTrue(appvm.property_is_default('audiovm'))
+        self.app.default_audiovm = audiovm
+        self.assertEventFired(holder, 'property-set:default_audiovm',
+                              kwargs={'name': 'default_audiovm',
+                                      'newvalue': 'sys-audio'})
+
+        self.assertIn('audiovm-sys-audio', appvm.tags)
```
Comment (marmarek) [inline qubes/tests/app.py]: Please also test if the tag is removed/changed correctly (setting audiovm to None, changing to other value).

File: qubes/tests/app.py [hunk 8]
```diff
@@ -596,18 +596,82 @@ class MyTestHolder(qubes.tests.TestEmitter, qubes.PropertyHolder):
...
+                                      'newvalue': 'sys-audio'})
+
+        # Set AudioVM
+        self.assertFalse(appvm.property_is_default('audiovm'))
+        appvm.audiovm = audiovm
+        self.assertIn('audiovm-sys-audio', appvm.tags)
+
+        # Change AudioVM
+        appvm.audiovm = guivm
+        self.assertIn('audiovm-sys-gui', appvm.tags)
```
Comment (marmarek) [inline qubes/tests/app.py]: Check also if `audiovm-sys-audio` tag is gone here.

File: qubes/tests/app.py [hunk 9]
```diff
@@ -596,18 +596,82 @@ class MyTestHolder(qubes.tests.TestEmitter, qubes.PropertyHolder):
...
+
+        # Empty AudioVM
+        del appvm.audiovm
+        self.assertNotIn('audiovm-', appvm.tags)
```
Comment (marmarek) [inline qubes/tests/app.py]: Sadly, it won't look for any `audiovm-*` tag. But IMO it's enough to add explicit `assertNotIn` for other two tags mentioned in the test. And leave this one here too, in case of weird errors.

File: qubes/tests/app.py [hunk 10]
```diff
@@ -596,18 +596,82 @@ class MyTestHolder(qubes.tests.TestEmitter, qubes.PropertyHolder):
...
+        appvm.guivm = guivm
         self.assertIn('guivm-sys-gui', appvm.tags)
 
+        # Change GuiVM
+        appvm.guivm = vncvm
+        self.assertIn('guivm-sys-vnc', appvm.tags)
+
+        # Empty GuiVM
+        del appvm.guivm
+        self.assertNotIn('guivm-', appvm.tags)
```
Comment (marmarek) [inline qubes/tests/app.py]: Same set of comments as for audiovm-* tag test below.

File: qubes/tests/app.py [hunk 11]
```diff
@@ -614,9 +614,12 @@ class MyTestHolder(qubes.tests.TestEmitter, qubes.PropertyHolder):
         # Change GuiVM
         appvm.guivm = vncvm
         self.assertIn('guivm-sys-vnc', appvm.tags)
+        self.assertNotIn('guivm-sys-vnc', appvm.tags)
```
Comment (marmarek) [inline qubes/tests/app.py]: Should be `guivm-sys-gui`...
